### DISCUSSION TITLE
Typescript

### DISCUSSION THREAD

**Discussion by jorenvandeweyer** [node-oauth/node-oauth2-server#269] [Ideas]
There has been a few times a suggestion to rewrite this library in TypeScript, and even a few times a PR was opened featuring a full rewrite, which we had to decline due to major changes in the code base. 

Since @jankapunkt started rewriting the library in ES6+ (#212) instead of ES5 (thank you for that! 🎉 ), I think we finally can start migrating to typescript without major changes to the codebase.

I think we first should decide how our tsconfig should look like before we start writing any code.

I think we can start from this `tsconfig.json` file

```json
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "ES2021",
    "baseUrl": ".",
    "outDir": "dist",
    "allowJs": true, // This should be changed to false afterwards
    "declaration": true,
    "sourceMap": true,
    "removeComments": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "skipLibCheck": true,
    "strictNullChecks": true,
    "noImplicitAny": false,
  },
  "include": ["lib/**/*", "test/**/*"],
  "exclude": ["node_modules"]
}
```

**Comment by shrihari-prakash** [comment]
Hello @jorenvandeweyer ,

I suggest to use also the path attribute so that imports can look very clean and verbose. Something like

```json
"paths": {
      "@grants/*": ["lib/grants/*"],
      "@models/*": ["lib/models/*"],
... and so on....
    }

```

What do you think?

**Reply by jorenvandeweyer** [reply]
Yes, good idea.

**Comment by jankapunkt** [comment]
From my end we can start with this and move to TS step by step. This is also a good way for me to get into TS. One note - we should keep the documentation examples in JS to stay as inclusive as possible.

**Comment by binajmen** [comment]
Don't ask me how I ended up here, but my 2 cents on the subject would be to consider JSDoc instead.

> TypeScript is great for application development, and it's getting better all the time. But for library development, JSDoc annotations in plain JavaScript seems like the way to go.

A library like Svelte decided to drop TS in favour of JSDoc. You still keep type checking (+ all the perks like documentation, etc.), plus it allows you to do things incrementally.

Just some food for thought ;)
